### PR TITLE
[test] add fuzz test for topk

### LINKED ISSUE #7749: TopK Fuzz Tests 🐝
### Is your feature request related to a problem or challenge?

After https://github.com/apache/arrow-datafusion/pull/7721 a `SortExec` with a limit will use a special `TopK` . We have basic unit tests, but I think the coverage could be improved, specifically with Fuzz testing

### Describe the solution you'd like

What I would like is a new fuzz test to be added to the the existing fuzz cases: https://github.com/apache/arrow-datafusion/tree/main/datafusion/core/tests/fuzz_cases

The structure of `SortTest` in https://github.com/apache/arrow-datafusion/blob/e95a24b5a260e0e2f603d52682d36cce192676f8/datafusion/core/tests/fuzz_cases/sort_fuzz.rs#L111 might be a good one to follow

The basic outline would be:
1. Create an input with several columns (integers, strings, floats)  
2. Reorder the input randomly
3. Divide the input up multiple batches using [`make_staggered_batches`](https://github.com/apache/arrow-datafusion/blob/0d82620bb45db35ce30ec170cb68e953fbc94366/datafusion/core/tests/fuzz_cases/aggregate_fuzz.rs#L172)
4. Run a query like `SELECT * FROM t ORDER BY <col(s)> LIMIT <N>` and collect the output
5. Compute the expected result programmatically (e.g. by sort the data, prior to creating RecordBatches)
6. Ensure the output matches the expected result

Input size: 1000 rows  

Parameters to vary
1. sort cols: (int), (string), (float), (int, string), (string, int), etc.
2. N: 1, 10, 100, 300 (aka how many are kept)

Bonus points
  make it easy to add new columns / types (e.g. like string dictionary)

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Tangruilin**
## Which issue does this PR close?
Closes #7749 .

## Rationale for this change
Add fuzz test for topk

## What changes are included in this PR?
1. This PR add i32 type fuzz test for topk;
2. This PR add fn to get f64 and string batched but now not use(for the fn batches_to_vec can only work for i32)

TODO: This PR is not completed. I need some suggestions that if i need to modify `fn batches_to_vec`

## Are these changes tested?
this PR is a test

## Are there any user-facing changes?
no

### CODE REVIEW

**Comment by Tangruilin** [comment]
@alamb Please review it.

**Comment by alamb** [comment]
Thank you @Tangruilin  -- I plan to review this tomorrow. I look forward to it!

**Review by alamb — COMMENTED** [review]
Thank you @Tangruilin 

> TODO: This PR is not completed. I need some suggestions that if i need to modify fn batches_to_vec

I am not quite sure what you mean here. I think if you can make a `TopKScenario` type structure, it maybe easier to figure out if you need to modify `batches_to_vec`

**Comment by alamb** [comment]
Mark as draft to signify this PR is not waiting on feedback anymore

File: datafusion/core/tests/fuzz_cases/sort_fuzz.rs [hunk 1]
```diff
@@ -22,89 +22,101 @@ use arrow::{
...
-    SortTest::new()
-        .with_int32_batches(1000000)
-        .with_pool_size(102400)
-        .with_should_spill(true)
-        .run()
-        .await;
+    for (batch_size, should_spill) in [(5, false), (20000, false), (1000000, true)] {
+        SortTest::new()
+            .with_int32_batches(batch_size)
+            .with_pool_size(102400)
```
Comment (Weijun-H) [inline datafusion/core/tests/fuzz_cases/sort_fuzz.rs:65]:
```suggestion
            .with_pool_size(100 * KB)
```

File: datafusion/core/tests/fuzz_cases/sort_fuzz.rs [hunk 2]
```diff
@@ -22,89 +22,101 @@ use arrow::{
...
+            .with_int32_batches(batch_size)
+            .with_pool_size(usize::MAX)
+            .with_should_spill(should_spill)
+            .run()
+            .await;
+    }
+}
+
+#[tokio::test]
+async fn test_sort_topk_i32() {
```
[the remark this answers is not in the document]
Comment (Tangruilin) [reply datafusion/core/tests/fuzz_cases/sort_fuzz.rs:85]: A good suggestion! I'll do it tomorrow and push it

**Comment by Tangruilin** [comment]
I run ci/scripts/rust_clippy.sh on my mac but get 
image
but the ci is 
image

the result is not some, I'm confused @alamb

**Comment by alamb** [comment]
> except that. I found that the rust should be stable —— run rustup update && rust default stable then it is solved. Are there docs for this. If not, maybe I can add it.

I think it is implicit in the [Testing setup:](https://arrow.apache.org/datafusion/contributor-guide/index.html#bootstrap-environment) part:

> rustup update stable DataFusion uses the latest stable release of rust


I agree this could be clearer in the documentation. Perhaps you can help here

**Comment by Tangruilin** [comment]
I have tried some ways, but there is not a prefect solution. I will update the PR tonight

**Comment by alamb** [comment]
Thanks @Tangruilin  -- I plan to review this more carefully tomorrow

**Comment by Tangruilin** [comment]
> Thanks @Tangruilin -- I plan to review this more carefully tomorrow

Don't forget this~~~

File: datafusion/core/tests/fuzz_cases/sort_fuzz.rs [hunk 3]
```diff
@@ -138,6 +154,44 @@ impl SortTest {
...
+
+        let df = ctx
+            .table(topk_scenario.table_name)
+            .await
+            .unwrap()
+            .sort(vec![
+                datafusion_expr::col(topk_scenario.col_name).sort(true, true)
+            ])
+            .unwrap()
+            .limit(0, Some(topk_scenario.limit))
```
Comment (alamb) [inline datafusion/core/tests/fuzz_cases/sort_fuzz.rs:189]: I verified the plan here has the expected `TopK` element:

```
| physical_plan                                              | GlobalLimitExec: skip=0, fetch=10                    |
|                                                            |   SortExec: TopK(fetch=10), expr=[x@0 ASC]           |
|                                                            |     MemoryExec: partitions=1, partition_sizes=[38]   |
```

**Review by alamb — APPROVED** [review]
Thank you very much @Tangruilin  and @Weijun-H  -- this very nicely done. I am sorry for my delay in reviewing. I think this PR adds additional coverage and is structured to make future improvements straghtforward.

🙏 great job

**Comment by alamb** [comment]
BTW I am working on an extesion of this test to support multiple columns

**Pull request merged**